### ISSUE TITLE
Please expand tksupport docs to explain how to make Mac Quit menu item work

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3776]
|<img alt="rowen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| rowen reported|
|-|-|
|Trac ID|trac#3776|
|Type|enhancement|
|Created|2009-04-15 21:09:35Z|

When using tksupport on MacOS X using Aqua Tcl/Tk the Quit menu item is unresponsive.

The fix is simple, and it would be very helpful to include it in the documentation for tksupport (or even apply it automatically as part of tksupport?):

root.createcommand("::tk::mac::Quit", reactor.stop)

This issue seems to be unique to MacOS X and the Aqua version of Tcl/Tk because the application automatically gets a standard Quit menu item, its normal behavior is useless when using Twisted and tksupport, and the fix is not at all obvious because the Quit menu item is not a standard tk menu item, so it cannot be reconfigured in the standard way.

Note: the above fix is needed in addition to the usual fix of catching WM_DELETE_WINDOW, which is needed on all platforms and is already in the Twisted documentation:

root.protocol('WM_DELETE_WINDOW', reactor.stop)

Attachments:

* [aqua-quit-3776.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0e9/0e960b3c0f76eb4b5e063abff708e0f9e4a1f25c/bbfe35769493d736d3b4e20c18b52859dcd476b8.patch) (577 bytes) - added by thijs on 2011-05-27 11:42:11Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3776 3776
type__enhancement enhancement
reporter__rowen rowen
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__tksupport_documentation tksupport documentation
time__1239829775000000 1239829775000000
changetime__1307311239000000 1307311239000000
version__None None
owner__thijs thijs
cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Thanks, for filing this detailed ticket, Rowen!

Is there an appropriate bit of Tkinter documentation that we can link to, in order to point users at something a bit more comprehensive rather than just giving them a single magic string?  It occurs to me that there may be other subtleties for integrating properly with other Tk platform features, and it would be good to have a detailed resource to link to rather than filing a ticket for each one individually.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

I can't see any reason this needs to be targeted at 9.0 (8.2+1).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to rowen|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra removed owner|
|-|-|

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess set owner to @thijstriemstra|
|-|-|

Thanks for working on this, thijs. In short, looks great, please merge!

Since the ticket is a few years old, I did check on my Mac running OS X 10.6 that this is still necessary, borrowing a code snippet from [ticket 1433](http://twistedmatrix.com/trac/ticket/1433):

```
from twisted.internet import reactor, tksupport, task
from Tkinter import *

root = None

def printstuff():
    print 'stuff'
task.LoopingCall(printstuff).start(1)

def window_closed(e=None):
    global root

    if root is not None:
        tksupport.uninstall()
        root.destroy()

root=Tk()
root.protocol("WM_DELETE_WINDOW", window_closed)
#root.createcommand("::tk::mac::Quit", reactor.stop)
tksupport.install(root)
reactor.run()
```

With the `createcommand` line commented out, after running the script if I try to quit with command-Q (or via the menu Python -> Quit Python), nothing happens, and with the line uncommented, quitting does work as desired.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [32048]) Apply aqua-quit-3776.patch: document a workaround for the Aqua Tcl/Tk Quit menu
on Mac OS X.

Author: thijs
Reviewer: jesstess
Fixes: [#3776](https://github.com/twisted/twisted/issues/3776)
